cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/documentstore/DatabaseImpl.java

```diff
@@ -805,9 +805,25 @@ boolean isOpen() {
         Misc.checkNotNull(revisions, "Input revisions");
         Misc.checkArgument(!revisions.isEmpty(), "revisions cannot be empty");
 
-        // TODO hoist down queue.submit to just surround the call to RevsDiffBatchCallable
         try {
-            return get(queue.submit(new RevsDiffCallable(revisions)));
+
+            ValueListMap<String, String> missingRevs = new ValueListMap<String, String>();
+
+            // Break down by docId first to avoid potential rev ID clashes between doc IDs
+            for (Map.Entry<String, List<String>> entry : revisions.entrySet()) {
+                String docId = entry.getKey();
+                List<String> revs = entry.getValue();
+                // Partition into batches to avoid exceeding placeholder limit
+                // The doc ID will use one placeholder, so use limit - 1 for the number of
+                // revs for the remaining placeholders.
+                List<List<String>> batches = CollectionUtils.partition(revs,
+                        SQLITE_QUERY_PLACEHOLDERS_LIMIT - 1);
+
+                for (List<String> revsBatch : batches) {
+                    missingRevs.addValuesToKey(docId, get(queue.submit(new RevsDiffBatchCallable(docId, revsBatch))));
```

Comment:
This actually introduces a subtle behaviour change. Currently all the batches are retrieved from the same queue operation. After this change the batches could be interleaved between other operations on the queue that have changed the state of the document store.

The impact of the change could be minimised by submitting all the operations, then getting all the results, rather than submitting/getting each batch before moving onto the next one, but there would still be no guarantees that other operations didn't happen between batches.

I'm not actually sure if this presents any problems or not, I was trying to think of a few scenarios:

1. A new local revision is written between revs diff batch calculations that matches a server side revision (in the case of a document with many revisions such that more than 1 batch exists for the same doc ID) then it is possible we would record a revision as missing when it was present and subsequently attempt to add it, which I believe could cause a replication to fail.

1. A new local revision is deleted between revs diff batch calculations. I think this has no impact currently, but may introduce additional complexity if we later introduced a purge.

Thoughts?